tensorflow/compiler/mlir/xla/transforms/lhlo_legalize_to_loop.cc

```diff
@@ -0,0 +1,397 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements logic for lowering LHLO dialect to Affine dialect.
+//
+#include <iostream>
+#include "absl/memory/memory.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"  // TF:llvm-project
+#include "mlir/Dialect/LoopOps/EDSC/Builders.h" 
+#include "mlir/IR/Attributes.h"  // TF:llvm-project
+#include "mlir/IR/Location.h"  // TF:llvm-project
+#include "mlir/IR/MLIRContext.h"  // TF:llvm-project
+#include "mlir/IR/PatternMatch.h"  // TF:llvm-project
+#include "mlir/IR/StandardTypes.h"  // TF:llvm-project
+#include "mlir/Pass/Pass.h"  // TF:llvm-project
+#include "tensorflow/compiler/mlir/xla/ir/lhlo_ops.h"
+
+namespace mlir {
+namespace xla_lhlo {
+
+namespace {
+
+constexpr StringRef kDhloMemrefFuncAttr = "xla_dhlo.memref_func";
+
+enum class ScheduleKind {
+  kLoop,
+  kRowReduction,
+  kColReduction,
+};
+
+std::vector<Operation*> getRootOps(FuncOp& func) {
+  std::vector<Operation*> root_ops;
+  auto arg_list = func.getArguments();
+  func.walk([&](Operation* op) {
+    if (op->getDialect() !=
+        op->getContext()->getRegisteredDialect("xla_lhlo")) {
+      return;
+    }
+    if (op->getParentOp() != func.getOperation()) {
+      return;
+    }
+    auto num_operands = op->getNumOperands();
+    if (num_operands == 0) return;
+    // for lhlo_op the last operand is always the result.
+    // TODO: any exceptions?
+    auto result = op->getOperand(num_operands - 1);
+    for (auto arg : arg_list) {
+      if (arg == result) {
+        root_ops.emplace_back(op);
+        return;
+      }
+    }
+  });
+  return root_ops;
+}
+
+// TODO: Only rank 2 to rank 1 reduction can be
+// supported for now
+bool isRowReduction(Operation* op) {
+  if (!isa<xla_lhlo::ReduceOp>(op)) {
+    return false;
+  }
+  auto reduce_op = cast<xla_lhlo::ReduceOp>(op);
+  if (reduce_op.dimensions().getNumElements() != 1) {
+    return false;
+  }
+  auto rank =
+      op->getOperand(0).getType().cast<MemRefType>().getRank();
+  auto dimensions = reduce_op.dimensions().getValues<int64_t>();
+  return ((*dimensions.begin() == 1) && (rank == 2));
```

Comment:
It looks that the above
"reduce_op.dimensions().getNumElements() != 1" check  has already ensure this is a X-to-1 reduction, 
so here the "(*dimensiond.begin()_ == 1" is to ensure this is a row X-to-1 reduction, correct?